File: src/main/java/ch/o2it/weblounge/kernel/webconsole/WebConsolePlugin.java
/*
 *  Weblounge: Web Content Management System
 *  Copyright (c) 2010 The Weblounge Team
 *  http://weblounge.o2it.ch
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this program; if not, write to the Free Software Foundation
 *  Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

package ch.o2it.weblounge.kernel.webconsole;

import ch.o2it.weblounge.common.site.Site;
import ch.o2it.weblounge.kernel.WebloungeSharedResources;

import org.apache.commons.io.IOUtils;
import org.apache.felix.webconsole.AbstractWebConsolePlugin;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 * This class implements plugin for the felix webconsole that provides access to
 * weblounge functionality as well as sites.
 */
public class WebConsolePlugin extends AbstractWebConsolePlugin {

  /** Serial version uid */
  private static final long serialVersionUID = 868658142169523757L;

  /** Plugin uri */
  private static final String URI = "weblounge";

  /** Plugin name */
  private static final String TITLE = "Weblounge";

  /** Path to the main index */
  private static final String INDEX = "/webconsole/index.html";

  /** Template key for the sites data */
  private static final String SITES_DATA_KEY = "sites-data";

  /** Template key for the total sites count */
  private static final String SITES_COUNT_KEY = "sites-count";

  /** Template key for the number of active sites */
  private static final String SITES_ACTIVE_COUNT_KEY = "sites-active";

  /** Template key for the number of inactive sites */
  private static final String SITES_INACTIVE_COUNT_KEY = "sites-inactive";

  /** Template key for the uri to the weblounge shared resources */
  private static final String SHARED_RESOURCES_KEY = "shared-resources";

  /** The list of registered sites */
  private List<Site> sites = new ArrayList<Site>();

  /** Reference to the shared resources */
  private WebloungeSharedResources sharedResources = null;

  /**
   * {@inheritDoc}
   * 
   * @see org.apache.felix.webconsole.AbstractWebConsolePlugin#getLabel()
   */
  @Override
  public String getLabel() {
    return URI;
  }

  /**
   * {@inheritDoc}
   * 
   * @see org.apache.felix.webconsole.AbstractWebConsolePlugin#getTitle()
   */
  @Override
  public String getTitle() {
    return TITLE;
  }

  /**
   * {@inheritDoc}
   * 
   * @see org.apache.felix.webconsole.AbstractWebConsolePlugin#renderContent(javax.servlet.http.HttpServletRequest,
   *      javax.servlet.http.HttpServletResponse)
   */
  @Override
  protected void renderContent(HttpServletRequest req, HttpServletResponse res)
      throws ServletException, IOException {
    InputStream is = null;
    try {
      is = getClass().getResourceAsStream(INDEX);
      String template = IOUtils.toString(is);

      // Process the template
      template = addSharedResourcesMountpoint(template);
      template = addSitesData(template);

      // Write the template to the output
      res.getWriter().print(template);
    } finally {
      IOUtils.closeQuietly(is);
    }
  }

  /**
   * Returns the replacement for <code>sites-data</code> in the template.
   * 
   * @param the
   *          template
   * @return the sites data
   */
  protected String addSharedResourcesMountpoint(String template) {
    String uri = null;
    if (sharedResources != null)
      uri = sharedResources.getResourcesMountpoint();
    else
      uri = WebloungeSharedResources.DEFAULT_RESOURCES_MOUNTPOINT;
    template = template.replace("${" + SHARED_RESOURCES_KEY + "}", uri);
    return template;
  }

  /**
   * Returns the replacement for <code>sites-data</code> in the template.
   * 
   * @param the
   *          template
   * @return the sites data
   */
  protected String addSitesData(String template) {
    int activeSites = 0;
    int inactiveSites = 0;

    // Collect and process the site data
    StringBuffer sitesData = new StringBuffer();
    synchronized (sites) {
      int i = 1;
      for (Site site : sites) {
        String state = site.isRunning() ? "active" : "inactive";
        activeSites += site.isRunning() ? 1 : 0;
        inactiveSites += !site.isRunning() ? 1 : 0;
        sitesData.append("{");
        sitesData.append("\"id\":").append(i).append(",");
        sitesData.append("\"name\":\"").append(site.getName()).append("\",");
        sitesData.append("\"version\":\"").append("-").append("\",");
        sitesData.append("\"status\":\"").append(state).append("\"");
        sitesData.append("}");
        i++;
      }
    }
    template = template.replace("${" + SITES_DATA_KEY + "}", sitesData.toString());

    // Total number of sites
    template = template.replace("${" + SITES_COUNT_KEY + "}", Integer.toString(sites.size()));

    // Number of active sites
    template = template.replace("${" + SITES_ACTIVE_COUNT_KEY + "}", Integer.toString(activeSites));

    // Number of inactive sites
    template = template.replace("${" + SITES_INACTIVE_COUNT_KEY + "}", Integer.toString(inactiveSites));

    return template;
  }

  /**
   * Adds a site to the list of registered sites.
   * 
   * @param site
   *          the site to add
   */
  void addSite(Site site) {
    synchronized (sites) {
      sites.add(site);
    }
  }

  /**
   * Removes a site from the list of registered sites.
   * 
   * @param site
   *          the site to remove
   */
  void removeSite(Site site) {
    synchronized (sites) {
      sites.remove(site);
    }
  }

  /**
   * Sets a reference to the shared resources.
   * 
   * @param sharedResources
   *          the shared resources
   */
  void setSharedResources(WebloungeSharedResources sharedResources) {
    this.sharedResources = sharedResources;
  }

}

File: modules/weblounge-kernel/src/main/resources/html/scripts/webconsole-plugin.js
/*
 *  Weblounge: Web Content Management System
 *  Copyright (c) 2010 The Weblounge Team
 *  http://entwinemedia.com/weblounge
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this program; if not, write to the Free Software Foundation
 *  Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

/**
 * Main rendering method that is called once the whole view has been loaded.
 */
function renderData(e)  {
	switch(e.status) {
		case  0: // no sites
			$(".statline").html(i18n.stat_no_sites);
			$('#scr').addClass('ui-helper-hidden');
			break;
		default:
			$(".statline").html(i18n.stat_ok.msgFormat(e.sites_count, e.sites_active, e.sites_inactive));
			$('#scr').removeClass('ui-helper-hidden');
	
			tableBody.empty();
			for (var i in e.sites_data) {
				site(e.sites_data[i]);
			}
			$("#plugin_table").trigger("update");
			
			//if (showSiteDetails) renderDetails(e);
			initStaticWidgets();
	}
}

/**
 * Processes a single site from the data that was loaded and adds it to the site
 * table.
 */
function site(/* Object */ site) {
	var id = site.id;
	var name = site.name;
	var tableEntry = tableEntryTemplate.clone().appendTo(tableBody).attr('id', 'entry' + id);

	tableEntry.find('.bIcon').attr('id', 'img' + id).click(function() {
		showDetails(id);
	}).after(showSiteDetails ? name : ('<a href="' + pluginRoot + '/' + id + '">' + name + '</a>'));

	tableEntry.find('td:eq(1)').text(site.state);

	// setup buttons
	if (site.stateRaw == 0) {
		tableEntry.find('li:eq(0)').removeClass('ui-helper-hidden').click(function() {
			setSiteStatus(id, 'enable')
		});
	} else {
		tableEntry.find('li:eq(1)').removeClass('ui-helper-hidden').click(function() {
			setSiteStatus(id, 'disable')
		});
	}
}

/**
 * Sets the site status of the indicated site according to the given status and
 * triggers a ui update.
 */
function setSiteStatus(/* long */ id, /* String */ status) {
	$.put(siteServiceUri + "/" + id, {"status":status}, function(data) {
		loadData();
	}, "xml");	
}

/**
 * Loads the details for the site with the given identifier and makes sure the
 * details are being rendered.
 */
function showDetails(/* long */ id) {
	$.get(pluginRoot + "/" + id + ".json", null, function(data) {
		renderDetails(data);
	}, "json");
}

/**
 * Loads all data and triggers rendering.
 */
function loadData() {
	$.get(pluginRoot + "/index.json", null, function(data) {
		renderData(data);
	}, "json");	
}

/**
 * Hides the site details.
 */
function hideDetails(/* String */ id) {
	var __test__ = $("#img" + id);
	$("#img" + id).each(function() {
		$("#pluginInlineDetails").remove();
		$(this).
			removeClass('ui-icon-triangle-1-w').// left
			removeClass('ui-icon-triangle-1-s').// down
			addClass('ui-icon-triangle-1-e').// right
		    attr("title", "Details").
			unbind('click').click(function() {showDetails(id)});
	});
}

/**
 * Adds the site details.
 */
function renderDetails(/* Object */ site) {
	site = site.site[0];
	var id = site.id
	$("#pluginInlineDetails").remove();
	var __test__ = $("#entry" + id);
	$("#entry" + id + " > td").eq(1).append("<div id='pluginInlineDetails'/>");
	$("#img" + id).each(function() {
		if (showSiteDetails) {
			var ref = window.location.pathname;
			ref = ref.substring(0, ref.lastIndexOf('/'));
			$(this).
				removeClass('ui-icon-triangle-1-e'). // right
				removeClass('ui-icon-triangle-1-s'). // down
				addClass('ui-icon-triangle-1-w'). // left
				attr("title", "Back").
				unbind('click').click(function() {window.location = ref});
		} else {
			$(this).
				removeClass('ui-icon-triangle-1-w'). // left
				removeClass('ui-icon-triangle-1-e'). // right
				addClass('ui-icon-triangle-1-s'). // down
				attr("title", "Hide Details").
				unbind('click').click(function() {hideDetails(id)});
		}
	});
	$("#pluginInlineDetails").append("<table border='0'><tbody></tbody></table>");
	var details = site.props;
	for (var idx in details) {
		var prop = details[idx];
		var key = i18n[prop.key] ? i18n[prop.key] : prop.key; // i18n

		var txt = "<tr><td class='aligntop' noWrap='true' style='border:0px none'>" + key + "</td><td class='aligntop' style='border:0px none'>";	
		if (prop.value) {
			if ($.isArray(prop.value)) {
				var i = 0;
				for(var pi in prop.value) {
					var value = prop.value[pi];
					if (i > 0) { txt = txt + "<br/>"; }
					var span;
					if (value.substring(0, 2) == "!!") {
						txt = txt + "<span style='color: red;'>" + value + "</span>";
					} else {
						txt = txt + value;
					}
					i++;
				}
			} else {
				txt = txt + prop.value;
			}
		} else {
			txt = txt + "\u00a0";
		}
		txt = txt + "</td></tr>";
		$("#pluginInlineDetails > table > tbody").append(txt);
	}
}

var tableBody = false;
var tableEntryTemplate = false;

/**
 * The document has been fully loaded by the client. Time to add more content
 * and register event handler.
 */
$(document).ready(function(){
	tableBody = $('#plugin_table tbody');
	tableEntryTemplate = tableBody.find('tr').clone();

	renderData(pluginData);

	$(".reloadButton").unbind('submit');
	$(".reloadButton").click(loadData);

	var extractMethod = function(node) {
		var link = node.getElementsByTagName("a");
		if (link && link.length == 1) {
			return link[0].innerHTML;
		}
		return node.innerHTML;
	};
	
	$("#plugin_table").tablesorter({
		headers: {
			0: { sorter:"digit"},
			2: { sorter: false }
		},
		sortList: [[0,1]],
		textExtraction:extractMethod
	});
});
